File: scripts/ingest.py
from server.data_loaders.gmail import read_emails, process_and_store_email
from server.data_loaders.files import process_files
from server.database.db import Database
from server.constants import folder_path
from sqlalchemy.exc import IntegrityError
from server.data_loaders.history import get_history, process_and_store_history

def ingest_recent_emails(days: int = 7):
    # Initialize the database
    db = Database()

    # Get recent email threads
    recent_email_threads = read_emails(days)

    # Process each email in each thread
    for thread in recent_email_threads:
        for thread_id, emails in thread.items():
            for email in emails:
                try:
                    stored_emails = process_and_store_email(email)
                    for stored_email in stored_emails:
                        print(f"Stored email chunk: {stored_email['id']} - {stored_email['subject']}")
                except IntegrityError:
                    print(f"Email already exists: {email.subject}")
                    db.session.rollback()
                except Exception as e:
                    print(f"Error processing email: {e}")
                    db.session.rollback()

    print("Email ingestion complete.")

def ingest_files(folder_path: str):
    db = Database()
    try:
        process_files(folder_path)
    except Exception as e:
        print(f"Error during file ingestion: {e}")
        db.session.rollback()

def ingest_browser_history():
    try:
        print("Getting history")
        history_entries = get_history()
        print(history_entries)
        process_and_store_history(history_entries)
    except Exception as e:
        print(f"Error during history ingestion: {e}")
        db.session.rollback()

if __name__ == "__main__":
    db = Database()
    
    # Ingest emails
    ingest_recent_emails()
    
    # Ingest files
    ingest_files(folder_path)

    # Ingest browser history
    ingest_browser_history()

File: server/src/scripts/ingest.ts
import { container } from '../utils/container';
import { PrismaClient } from '@prisma/client';
import { gmailClient } from '../services/clients/gmail';
import { processAndStoreEmail } from '../services/dataLoaders/email';
import { processFiles } from '../services/dataLoaders/files';
import { getContentsForUrl } from '../services/clients/exa';
import { createLink } from '../dao/linkDao';
import dotenv from 'dotenv';
import fs from 'fs';
import path from 'path';

dotenv.config();

const FOLDER_PATH = process.env.FOLDER_PATH || 'jeff_storage';

async function ingestRecentEmails(days: number = 7) {
  const prisma = container.get(PrismaClient);

  try {
    const recentEmailThreads = await gmailClient.readEmails(days);

    for (const thread of Object.values(recentEmailThreads)) {
      for (const email of thread) {
        try {
          const storedEmail = await processAndStoreEmail(email);
          console.log(`Stored email: ${storedEmail.id} - ${storedEmail.subject}`);
        } catch (error: any) {
          if (error.code === 'P2002') {
            console.log(`Email already exists: ${email.subject}`);
          } else {
            console.error(`Error processing email: ${error}`);
          }
        }
      }
    }
  } catch (error) {
    console.error('Error during email ingestion:', error);
  }

  console.log("Email ingestion complete.");
}

async function ingestFiles(folderPath: string) {
  try {
    await processFiles(folderPath);
  } catch (error) {
    console.error('Error during file ingestion:', error);
  }
}

async function ingestBrowserHistory() {
  try {
    console.log("Getting history");
    const historyFilePath = path.join(FOLDER_PATH, 'browser_history.json');
    const historyData = fs.readFileSync(historyFilePath, 'utf8');
    const historyEntries = JSON.parse(historyData);

    for (const entry of historyEntries) {
      try {
        const content = await getContentsForUrl(entry.url);
        await createLink({
          url: entry.url,
          title: entry.title,
          content: content,
        });
        console.log(`Stored link: ${entry.url}`);
      } catch (error) {
        console.error(`Error processing link ${entry.url}:`, error);
      }
    }
  } catch (error) {
    console.error('Error during history ingestion:', error);
  }
}

async function main() {
  const prisma = container.get(PrismaClient);

  // Ingest emails
  await ingestRecentEmails();

  // Ingest files
  await ingestFiles(FOLDER_PATH);

  // Ingest browser history
  await ingestBrowserHistory();

  await prisma.$disconnect();
}

main().catch((error) => {
  console.error('Error in main ingest process:', error);
  process.exit(1);
});